Add blacksmith tooltip to ForgingBonus in tfc 1.20.x

koenigstag opened the pull request:
Small changes to add author NBT to ForgingBonus capability in order to display blacksmith nickname in item tooltip.
Added translation to all language files.

koenigstag committed:
Add blacksmith tooltip to ForgingBonus

koenigstag commented:
Blacksmith tooltip in action
![image](https://github.com/user-attachments/assets/cd8ef3ff-21d9-48e7-abf0-40fdaf3d885d)

alcatrazEscapee commented:
First, do you speak all the languages you have added translations for? I'm assuming not, in which case you should not have added them.
Second, this was already added in fd03de3771ba099e5e35cc469c4d28845ff4b6c3 in 1.21.x, it should be ported without changes, which this notably wasn't (accepting for API differences).

koenigstag commented:
I can say that I've translated 4 languages correctly cause I used to learn and use them (Ru, Uk, De, Pl), other 3 have single translation possible (Es, Pt, Tr), and for other asian languages I've used google translate for single word translation, which should be also correct.

koenigstag commented:
Will work on porting this to 1.20.x

@alcatrazEscapee 
Upd. Sadly, Forge 1.20.1 is missing features from 1.21 such as DataComponent, Registries.DATA_COMPONENT_TYPE, and also there is no such package as net.minecraft.network.codec.

So will search for a workaround if any.
Or will try to use old-fashioned NBT system to not break compatibility with mc 1.20.1.

eerussianguy commented:
You would need to do it with NBT, yes

koenigstag commented:
Its already done via NBT author tag in this MR

`stack.getOrCreateTag().putString(KEY_AUTHOR, player.getName().getString());`

koenigstag committed:
Inherit method logic

alcatrazEscapee commented:
Both issues I pointed out are still present:
- The implementation is different in 1.21 than it is here. To be specific, this is missing implementation for copying, has a different display than 1.21, has issues with empty string handling and is in general, not acceptable.
- No, google translated entries are **not** acceptable from someone who doesn't speak the language. Nor is "this should be correct"

koenigstag commented:
What should I do with Wielding Recipe, as the logic there requires a new "Behavior bonus" field ?
The rewrite of recipe will break other add-on recipes.
I think I can add a new field and not remove the old one. The logic can be combined and forgiving as for migration period.

koenigstag committed:
Refactor author tooltip for similarity with 1.21

koenigstag committed:
Prioritize "behavior" over the "combine bonus" logic

koenigstag commented:
- Added logic to copy bonus in WeldingRecipe and CopyForgingBonusModifier classes
- Added logic for new "bonus behavior" field in WeldingRecipe. Bonus behavior logic has the priority over "combine bonus".

Is there anything else I should fix?
